pkg/backend/filestate/store.go

```diff
@@ -0,0 +1,149 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/encoding"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/fsutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+// These should be constants
+// but we can't make a constant from filepath.Join.
+var (
+	// StacksDir is a path under the state's root directory
+	// where the filestate backend stores stack information.
+	StacksDir = filepath.Join(workspace.BookkeepingDir, workspace.StackDir)
+
+	// HistoriesDir is a path under the state's root directory
+	// where the filestate backend stores histories for all stacks.
+	HistoriesDir = filepath.Join(workspace.BookkeepingDir, workspace.HistoryDir)
+
+	// BackupsDir is a path under the state's root directory
+	// where the filestate backend stores backups of stacks.
+	BackupsDir = filepath.Join(workspace.BookkeepingDir, workspace.BackupDir)
+)
+
+// referenceStore stores and provides access to stack information.
+//
+// Each implementation of referenceStore is a different version of the stack
+// storage format.
+type referenceStore interface {
+	// StackBasePath returns the base path to for the file
+	// where snapshots of this stack are stored.
+	//
+	// This must be under StacksDir.
+	//
+	// This is the path to the file without the extension.
+	// The real file path is StackBasePath + ".json"
+	// or StackBasePath + ".json.gz".
+	StackBasePath(*localBackendReference) string
+
+	// HistoryDir returns the path to the directory
+	// where history for this stack is stored.
+	//
+	// This must be under HistoriesDir.
+	HistoryDir(*localBackendReference) string
+
+	// BackupDir returns the path to the directory
+	// where backups for this stack are stored.
+	//
+	// This must be under BackupsDir.
+	BackupDir(*localBackendReference) string
+
+	// ListReferences lists all stack references in the store.
+	ListReferences() ([]*localBackendReference, error)
+
+	// ParseReference parses a localBackendReference from a string.
+	ParseReference(ref string) (*localBackendReference, error)
+}
+
+// legacyReferenceStore is a referenceStore that stores stack
+// information with the legacy layout that did not support projects.
+//
+// This is the format we used before we introduced versioning.
+type legacyReferenceStore struct {
+	b *localBackend
```

Comment:
Rather than depending on the whole backend struct here could we just take the bucket? Is there some layering of these types such that we don't have any cycles or duplicated fields? 

Like it feels like the ideal construction would be something like:
1. Create a bucket
2. Use that to create a store
3. Use that to create a backend
Each one having everything it needs to do it's part of the work and not needing to call upwards into the higher layers.

If that's an easy refactor lets do it, if that's complicated than I think this is ok for now. The whole backend codes gonna need an overhaul at some point anyway.

Reply:
Indeed!
This is something I wanted to do (the cyclic dependency between store and backend grated at me a bit) but originally deferred it to a future follow up in my head.
I think now's a good time to break that if possible. 
We really just need two pieces in the store: bucket and a way to access the "current project" so that the localBackendReference can look it up. 
I'll do the refactor here; it'll also simplify the tests because ideally store tests shouldn't have to know about the Pulumi.yaml.